File: web/src/bundle/manifest.ts
/**
 * T053 — the launch-time integrity check, and the canonicalization it rests on.
 *
 * ## The digest rule, restated because getting it slightly wrong is undiagnosable
 *
 * `integrity.manifest_sha256` is the SHA-256 of the manifest serialized per
 * **RFC 8785 (JSON Canonicalization Scheme)** with the **`integrity` key removed**
 * before serialization — *removed*, not `null`, not `""`. Lowercase hex.
 *
 * The writer is Python (`rfc8785`) and this verifier is TypeScript (`canonicalize`,
 * Apache-2.0, pinned per ADR-0007). They must agree byte-for-byte, and by default
 * they do **not** — `docs/data/bundle-manifest.md` and ADR-0025 A5 record the two
 * divergences that a hand-rolled `JSON.stringify` with sorted keys walks straight into:
 *
 *   - **escaping**, which fires on *every* manifest: `json.dumps` defaults to
 *     `ensure_ascii=True` and escapes the `©` that every ODbL attribution carries;
 *   - **number formatting**, which fires *intermittently*: Python renders `28.0`
 *     where JavaScript renders `28`, but only when a bbox bound lands on a whole
 *     degree — so it breaks for some areas and not others, which reads like
 *     corruption rather than like a serialization bug.
 *
 * A mismatch surfaces **offline, on the traveller's device, with no way to diagnose
 * it**. So: no hand-rolling. RFC 8785 or nothing.
 *
 * ## Why the digest is computed over the RAW parsed JSON, never over our own type
 *
 * {@link parseManifest} narrows an unknown blob to {@link BundleManifestV1} and drops
 * what it does not model. Hashing *that* would be a bug with a long fuse: an M2 bundle
 * carrying `schematic` (a field this client has no reason to model yet) would fail its
 * own integrity check on an M1 client, and the failure would look like corruption.
 * {@link manifestDigest} therefore takes the **raw** parsed value. The narrowed type is
 * for rendering; the raw value is for hashing. They are not interchangeable.
 */

import canonicalize from 'canonicalize'

import {
  WITHHELD_REASONS,
  type BundleManifestV1,
  type HashedArtifact,
  type TileSourceV1,
  type WithheldEntry,
  type WithheldReason,
} from './types'

/** Lowercase hex SHA-256 of raw bytes — the card's hash format for every artifact. */
export async function sha256Hex(bytes: BufferSource): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', bytes)
  return [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, '0')).join('')
}

/**
 * The exact bytes the manifest digest is taken over: RFC 8785 canonical JSON of the
 * manifest **with `integrity` removed** (a hash cannot cover itself).
 *
 * Exported so a test can assert the *rule* — "the `integrity` key does not appear" —
 * rather than a literal string this implementation happens to emit.
 */
export function canonicalManifestBytes(raw: unknown): Uint8Array {
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
    throw new TypeError('manifest is not a JSON object')
  }
  // Shallow copy + `delete`: the key must be ABSENT, not nulled or emptied. A
  // `{...raw, integrity: undefined}` spread would also work in JS but survives as an
  // own key under some serializers — `delete` is unambiguous.
  const withoutIntegrity: Record<string, unknown> = { ...(raw as Record<string, unknown>) }
  delete withoutIntegrity.integrity

  const canonical = canonicalize(withoutIntegrity)
  if (canonical === undefined) throw new TypeError('manifest is not canonicalizable JSON')
  return new TextEncoder().encode(canonical)
}

/** SHA-256 of {@link canonicalManifestBytes} — what `integrity.manifest_sha256` holds. */
export async function manifestDigest(raw: unknown): Promise<string> {
  return sha256Hex(canonicalManifestBytes(raw))
}

/**
 * Does the manifest's own `integrity.manifest_sha256` match what the bytes say?
 *
 * Case-insensitive on the *claimed* side only: the card fixes the digest as lowercase
 * hex, and {@link sha256Hex} emits lowercase, so a manifest that shouts its hash is
 * accepted rather than reported as corrupt. Nothing else about the value is repaired.
 */
export async function verifyManifestDigest(raw: unknown): Promise<boolean> {
  const claimed = (raw as { integrity?: { manifest_sha256?: unknown } } | null)?.integrity
    ?.manifest_sha256
  if (typeof claimed !== 'string' || claimed.trim() === '') return false
  return claimed.trim().toLowerCase() === (await manifestDigest(raw))
}

/* ------------------------------------------------------- narrowing ---------- */

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v)
}

function str(v: unknown): string | null {
  return typeof v === 'string' && v.trim() !== '' ? v : null
}

function num(v: unknown): number | null {
  return typeof v === 'number' && Number.isFinite(v) ? v : null
}

function hashedArtifact(v: unknown, pathKey = 'path', hashKey = 'sha256'): HashedArtifact | null {
  if (!isRecord(v)) return null
  const path = str(v[pathKey])
  const sha256 = str(v[hashKey])
  return path && sha256 ? { path, sha256: sha256.toLowerCase() } : null
}

/**
 * EPSG:4326 `[west, south, east, north]` — the card's CRS discipline.
 *
 * Two things are checkable and both are refused rather than corrected: a bound outside
 * the coordinate ranges, and a mis-ordered extent (`west > east`, `south > north`).
 *
 * **What is deliberately NOT claimed: this does not detect a lon/lat swap.** For most
 * inhabited places both values fall inside ±90, so a swapped `[36.44, 28.0, …]` is
 * indistinguishable from a real bbox by inspection alone. The defence against a swap is
 * upstream — coordinates are computed in PostGIS/shapely and never emitted by a model
 * (AGENTS.md geo rules) — not here. Saying so keeps a future reader from trusting a
 * guard that does not exist.
 */
function bbox(v: unknown): readonly [number, number, number, number] | null {
  if (!Array.isArray(v) || v.length !== 4) return null
  const out = v.map(num)
  if (out.some((n) => n === null)) return null
  const [w, s, e, n] = out as [number, number, number, number]
  if (w < -180 || w > 180 || e < -180 || e > 180) return null
  if (s < -90 || s > 90 || n < -90 || n > 90) return null
  if (w > e || s > n) return null
  return [w, s, e, n]
}

function tileSource(v: unknown): TileSourceV1 | null {
  if (!isRecord(v)) return null
  const path = str(v.path)
  const sha256 = str(v.sha256)
  const box = bbox(v.bbox)
  const maxzoom = num(v.maxzoom)
  // The card's required subset. Anything short of it cannot render a basemap at all.
  if (!path || !sha256 || !box || maxzoom === null) return null
  return {
    path,
    sha256: sha256.toLowerCase(),
    bbox: box,
    maxzoom,
    ...(num(v.minzoom) !== null ? { minzoom: num(v.minzoom) as number } : {}),
    ...(str(v.format) ? { format: str(v.format) as string } : {}),
    ...(str(v.schema_ver) ? { schema_ver: str(v.schema_ver) as string } : {}),
    ...(str(v.build_source) ? { build_source: str(v.build_source) as string } : {}),
    ...(str(v.build_date) ? { build_date: str(v.build_date) as string } : {}),
    ...(str(v.tile_license) ? { tile_license: str(v.tile_license) as string } : {}),
    ...(str(v.attribution) ? { attribution: str(v.attribution) as string } : {}),
    ...(hashedArtifact(v.style) ? { style: hashedArtifact(v.style) as HashedArtifact } : {}),
    ...(isRecord(v.glyphs) && str(v.glyphs.path)
      ? {
          glyphs: {
            path: str(v.glyphs.path) as string,
            ...(str(v.glyphs.license) ? { license: str(v.glyphs.license) as string } : {}),
          },
        }
      : {}),
  }
}

function isWithheldReason(v: unknown): v is WithheldReason {
  return typeof v === 'string' && (WITHHELD_REASONS as readonly string[]).includes(v)
}

/**
 * Narrow the `withheld` list.
 *
 * An entry whose `reason` is not in the closed enum is **dropped, not coerced**. The
 * enum is closed precisely so `withheld` cannot become a free-text channel out of a
 * downloadable artifact (ADR-0025 A3); mapping an unknown reason onto a known one would
 * hand that channel back. A dropped entry costs one "needs connectivity" affordance,
 * which is the safe direction — the value is absent either way.
 */
function withheld(v: unknown): readonly WithheldEntry[] {
  if (!Array.isArray(v)) return []
  const out: WithheldEntry[] = []
  for (const entry of v) {
    if (!isRecord(entry)) continue
    const site_id = str(entry.site_id)
    const field = str(entry.field)
    if (site_id && field && isWithheldReason(entry.reason)) {
      out.push({ site_id, field, reason: entry.reason })
    }
  }
  return out
}

/**
 * Narrow an unknown blob to a `BundleManifestV1`, or `null` if it is not one.
 *
 * **Fail closed.** Every field the travel UI needs to resolve a path is required; a
 * manifest missing any of them yields `null`, and {@link ./index!openBundle} reports the
 * bundle unusable rather than rendering the part of the day that happened to survive
 * (FR-020). `textLicense` is the one field that is legitimately `null` — ADR-0024 makes
 * it `null` exactly when `content.narrations` is empty — so a *missing* `textLicense`
 * key and an explicit `null` both read as "no bundled narration text", and neither is a
 * reason to refuse the bundle.
 */
export function parseManifest(raw: unknown): BundleManifestV1 | null {
  if (!isRecord(raw)) return null
  if (raw.schema_ver !== 'BundleManifestV1') return null

  const bundle_id = str(raw.bundle_id)
  const itinerary_id = str(raw.itinerary_id)
  const created_at = str(raw.created_at)
  const size_bytes = num(raw.size_bytes)
  const pmtiles = isRecord(raw.tiles) ? tileSource(raw.tiles.pmtiles) : null
  const attribution = hashedArtifact(raw.attribution)
  const manifest_sha256 = isRecord(raw.integrity) ? str(raw.integrity.manifest_sha256) : null

  const r = isRecord(raw.routing) ? raw.routing : {}
  const walk_graph = str(r.walk_graph)
  const walk_graph_sha256 = str(r.walk_graph_sha256)
  const legs = str(r.legs)
  const legs_sha256 = str(r.legs_sha256)

  const c = isRecord(raw.content) ? raw.content : {}
  const sites = str(c.sites)
  const sites_sha256 = str(c.sites_sha256)
  const narrations = str(c.narrations)
  const narrations_sha256 = str(c.narrations_sha256)
  const itinerary = str(c.itinerary)
  const itinerary_sha256 = str(c.itinerary_sha256)

  if (
    !bundle_id ||
    !itinerary_id ||
    !created_at ||
    size_bytes === null ||
    !pmtiles ||
    !attribution ||
    !manifest_sha256 ||
    !walk_graph ||
    !walk_graph_sha256 ||
    !legs ||
    !legs_sha256 ||
    !sites ||
    !sites_sha256 ||
    !narrations ||
    !narrations_sha256 ||
    !itinerary ||
    !itinerary_sha256
  ) {
    return null
  }

  return {
    bundle_id,
    itinerary_id,
    created_at,
    size_bytes,
    schema_ver: 'BundleManifestV1',
    tiles: { pmtiles },
    routing: {
      walk_graph,
      walk_graph_sha256: walk_graph_sha256.toLowerCase(),
      legs,
      legs_sha256: legs_sha256.toLowerCase(),
    },
    content: {
      sites,
      sites_sha256: sites_sha256.toLowerCase(),
      narrations,
      narrations_sha256: narrations_sha256.toLowerCase(),
      itinerary,
      itinerary_sha256: itinerary_sha256.toLowerCase(),
    },
    attribution,
    textLicense: str(raw.textLicense),
    withheld: withheld(raw.withheld),
    integrity: { manifest_sha256: manifest_sha256.toLowerCase() },
  }
}

File: compiler/manifest.py
"""Manifest assembly — one hash per artifact, and the one canonicalization the seal is over.

T040/T040a: the last two stages of the compile order (`compiler/AGENTS.md`, tech-design
§5.3). Everything upstream has produced bytes; this module records **where** those bytes
live inside the bundle, **what** each one hashes to, and seals the whole into a
:class:`~commons.models.BundleManifestV1` a device can re-check offline. Field-level ground
truth is [`docs/data/bundle-manifest.md`](../docs/data/bundle-manifest.md); where anything
here reads differently, the card wins.

## What this module deliberately does NOT implement: the canonicalizer

``integrity.manifest_sha256`` is **RFC 8785 (JCS)** over the manifest with the ``integrity``
key **removed** — not nulled, not emptied — and both halves of that already exist, in two
places that must agree byte-for-byte:

* **Python, the writer**: :meth:`BundleManifestV1.canonical_payload` deletes the key and
  :meth:`BundleManifestV1.seal` digests the result through the pinned ``rfc8785``
  (Trail of Bits, Apache-2.0, ``~=0.1.4`` — a 0.x minor bump changes every bundle's hash);
* **TypeScript, the launch-time verifier** (T053): ``web/src/bundle/manifest.ts`` deletes
  the same key and canonicalizes through the pinned ``canonicalize`` npm package.

A third implementation here is exactly the divergence the named-standard rule exists to
prevent, so :func:`build_manifest` calls ``seal`` and nothing below hand-rolls a serializer.
Measured on the ``©``-plus-integral-bbox shape **every** real manifest carries:

    json.dumps(sort_keys=True)  {"attribution":"\\u00a9 OpenStreetMap …","bbox":[28.0,36.0,…]}
    rfc8785.dumps               {"attribution":"© OpenStreetMap …","bbox":[28,36,…]}

Both divergences are in that one line: ``json.dumps`` defaults to ``ensure_ascii=True`` and
escapes the ``©`` of the ODbL attribution (fires on every bundle), and it renders ``28.0``
where ECMAScript ``Number::toString`` renders ``28`` (fires only when a bbox bound lands on
a whole degree — intermittent, so it reads like corruption rather than like a serialization
bug). JCS pins string escaping and number formatting as well as key order; "sorted keys,
UTF-8, no whitespace" pins neither. A manifest that gets this wrong fails its own launch
check **offline, on the traveller's device, with no way to diagnose it**.

That the two sides actually agree is a measured fact, not an inference from the RFC
(2026-08-09): a manifest carrying the ODbL ``©`` and ``bbox: [28.0, 36.0, 28.5, 36.5]``
canonicalizes to the same 1,607 bytes and the same digest ``698bc6ed…`` under ``rfc8785``
and under ``canonicalize@3.0.0`` driven through ``manifest.ts``'s own copy-delete-
canonicalize path. Re-measure it, do not re-reason about it, if either pin moves.

## One hash per artifact, over raw bytes

Every path in the manifest has exactly one SHA-256 beside it, lowercase hex, taken over the
artifact's **raw bytes** — a hash spanning two files could not say which one corrupted.
:class:`Artifact` is that pairing, and :meth:`Artifact.from_file` streams so the PMTiles
archive (tens to hundreds of MB) is hashed without being held in memory.

## The frozen itinerary is a *different document* from the approved one

``content.itinerary_sha256`` hashes the bytes of :attr:`FrozenItinerary.data`; the approval
hash on the ``user_plan`` row is :meth:`ItineraryV1.canonical_sha256`, over the itinerary
alone. The two digests differ **by construction** — the frozen artifact carries the local
frame the plan row does not — and neither is a check on the other. Comparing them (or
"fixing" one to match) would break either the approval compare-and-set (ADR-0023) or the
bundle's own integrity check.

## T040a — the local frame travels with the day, or the day means nothing

Every bundled time is **area-local wall clock**, and what turns ``10:00`` into an instant is
``ItineraryV1.date`` plus the *area row's* ``timezone``/``country_code``. A database row is
not a bundle artifact. Offline the travel UI can read only the device clock — on a phone
still set to home time that is hours off, and the UI highlights the wrong stop with nothing
to detect the error against. So :func:`freeze_itinerary` copies both values into the frozen
artifact, top-level, **alongside** ``date``: same fix ADR-0025 ruling 1 applied to the
itinerary itself, one level down.

They are written as sibling keys rather than a nested object because the card's rule is that
``ItineraryV1`` *never restates* the frame — so there is no field for them to nest under, and
"alongside `date`" is where a reader looking for the day's frame will look. The consequence
is stated rather than discovered: **the artifact is not a valid ``ItineraryV1`` payload.**
``StampedModel`` is ``extra="forbid"``, so ``ItineraryV1.model_validate(json.loads(data))``
*fails* on it. Read it back with :func:`read_frozen_itinerary`, which separates the two
halves. The TypeScript reader (``web/src/travel/parse.ts``) ignores keys it does not model,
so the artifact narrows there unchanged.
"""

from __future__ import annotations

import hashlib
import json
from collections.abc import Sequence
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Any, Final
from uuid import UUID

import rfc8785

from commons.frame import LocalFrame
from commons.models import (
    ArtifactRef,
    BundleContent,
    BundleManifestV1,
    BundleRouting,
    BundleTiles,
    ItineraryV1,
    TileSourceV1,
    WithheldValue,
)

__all__ = [
    "FRAME_KEYS",
    "ITINERARY_PATH",
    "Artifact",
    "FrozenItinerary",
    "build_manifest",
    "canonical_manifest_bytes",
    "freeze_itinerary",
    "manifest_json_bytes",
    "read_frozen_itinerary",
]

#: Read size for :meth:`Artifact.from_file`. 1 MiB keeps a 200 MB PMTiles archive off the
#: heap while staying far above the syscall-per-byte end of the tradeoff.
_HASH_CHUNK: Final[int] = 1 << 20

#: The two keys T040a adds to the frozen itinerary artifact — the area's local frame, copied
#: from the ``area`` row at compile time. Exported so a reader can strip exactly these and no
#: others; they are the *only* keys in the artifact that are not ``ItineraryV1`` fields.
FRAME_KEYS: Final[tuple[str, str]] = ("timezone", "country_code")

#: Default bundle-relative path of the frozen itinerary (the card's example layout). A
#: default, not a constraint: the manifest records whatever path the artifact was given.
ITINERARY_PATH: Final[str] = "content/itinerary.json"


@dataclass(frozen=True, slots=True)
class Artifact:
    """One bundled file as the manifest records it: where it is, what it hashes to, how big.

    ``sha256`` is lowercase hex over the file's raw bytes — lowercase because
    :data:`~commons.models.Sha256Hex` refuses anything else, the casing being part of the
    contract with a TypeScript verifier that compares these as plain strings.
    ``hashlib.hexdigest()`` is lowercase by definition, so nothing here normalizes it.
    """

    #: Bundle-relative path; validated as a :data:`~commons.models.BundlePath` when the
    #: manifest is built, so an absolute or ``..``-bearing path fails there, not silently.
    path: str
    sha256: str
    #: Byte length of the artifact — summed into ``BundleManifestV1.size_bytes``, which FR-014
    #: reports *before* the download starts.
    size_bytes: int

    @classmethod
    def from_bytes(cls, path: str, data: bytes) -> Artifact:
        """Hash an in-memory artifact (every JSON/Markdown artifact the compiler generates)."""
        return cls(path=path, sha256=hashlib.sha256(data).hexdigest(), size_bytes=len(data))

    @classmethod
    def from_file(cls, path: str, source: Path) -> Artifact:
        """Hash an artifact on disk **without reading it into memory**.

        For the PMTiles archive, which is the one bundled artifact big enough for the
        difference to matter (the ≤200 MB budget is almost entirely tiles). ``source`` is a
        filesystem path; ``path`` is where the artifact will sit *inside the bundle*, and the
        two are unrelated on purpose — nothing about our staging layout should leak into the
        manifest a device reads.
        """
        digest = hashlib.sha256()
        size = 0
        with source.open("rb") as handle:
            while chunk := handle.read(_HASH_CHUNK):
                digest.update(chunk)
                size += len(chunk)
        return cls(path=path, sha256=digest.hexdigest(), size_bytes=size)


@dataclass(frozen=True, slots=True)
class FrozenItinerary:
    """The day, serialized for the bundle: the bytes to write, and how the manifest names them.

    Carries ``itinerary_id`` so :func:`build_manifest` reads the id off the same object it
    hashes. Passing the id separately would let ``itinerary_id`` and ``content.itinerary``
    name different days, and nothing downstream re-checks that pairing.
    """

    itinerary_id: UUID
    #: Exactly the bytes to write at ``artifact.path`` — hashing anything else breaks the
    #: check this whole module exists to make possible.
    data: bytes
    artifact: Artifact


def freeze_itinerary(
    itinerary: ItineraryV1, frame: LocalFrame, *, path: str = ITINERARY_PATH
) -> FrozenItinerary:
    """Freeze the planned day into the bundle's ``content.itinerary`` artifact (T040/T040a).

    The document is the itinerary's JSON dump plus the area's :data:`FRAME_KEYS` — see the
    module docstring for why the frame has to travel with the day and why it sits top-level.

    Serialized with **JCS**, the same canonicalization the manifest seal uses, though nothing
    requires it of an artifact: it makes recompiling an unchanged day byte-identical, so an
    unchanged ``content.itinerary_sha256`` means "the day did not change" rather than "the
    day did not change *and* the serializer felt the same way about dict order today".
    """
    payload = itinerary.model_dump(mode="json")

    # A frame key already present would mean `ItineraryV1` grew a field this function then
    # silently overwrote — the card says the itinerary never restates the frame, so if that
    # ever changes it is a schema decision, not something to paper over here.
    collisions = [key for key in FRAME_KEYS if key in payload]
    if collisions:
        raise ValueError(
            f"ItineraryV1 now carries {', '.join(collisions)} itself; the frozen artifact "
            "would overwrite it. Reconcile with docs/data/itinerary.md before compiling."
        )
    payload["timezone"] = frame.timezone
    payload["country_code"] = frame.country_code

    data = rfc8785.dumps(payload)
    return FrozenItinerary(
        itinerary_id=itinerary.id, data=data, artifact=Artifact.from_bytes(path, data)
    )


def read_frozen_itinerary(data: bytes) -> tuple[ItineraryV1, LocalFrame]:
    """Read a frozen artifact back into the day and the frame it is read in.

    The inverse of :func:`freeze_itinerary`, and the only correct way to validate the artifact
    in Python: the frame keys must come off *before* ``ItineraryV1.model_validate``, which is
    ``extra="forbid"`` and would otherwise refuse the compiler's own output.

    Raises ``ValueError`` if the frame is absent or not a pair of non-empty strings. Absent is
    refused rather than defaulted, for the reason :class:`~commons.frame.FrameUnresolved`
    exists: a day whose wall clock is unframed cannot be scheduled *or* checked, and the
    device has no second source to fall back on.
    """
    payload: Any = json.loads(data)
    if not isinstance(payload, dict):
        raise ValueError(f"a frozen itinerary is a JSON object, got {type(payload).__name__}")

    frame_values = {key: payload.pop(key, None) for key in FRAME_KEYS}
    missing = [
        key for key, value in frame_values.items() if not isinstance(value, str) or not value
    ]
    if missing:
        raise ValueError(
            f"frozen itinerary is missing its local frame ({', '.join(missing)}); every "
            "bundled time is area-local wall clock and is unreadable without it (T040a)"
        )

    return ItineraryV1.model_validate(payload), LocalFrame(
        timezone=str(frame_values["timezone"]), country_code=str(frame_values["country_code"])
    )


def canonical_manifest_bytes(manifest: BundleManifestV1) -> bytes:
    """The exact bytes ``integrity.manifest_sha256`` is taken over — JCS, ``integrity`` gone.

    A two-call convenience over the model's own :meth:`BundleManifestV1.canonical_payload`
    (which is what removes the key) and the pinned ``rfc8785`` — *not* a second canonicalizer.
    It exists because commons exposes the digest but not the bytes, and the bytes are what a
    cross-language comparison against ``web/src/bundle/manifest.ts`` has to be made over.
    """
    return rfc8785.dumps(manifest.canonical_payload())


def manifest_json_bytes(manifest: BundleManifestV1) -> bytes:
    """The bytes to **ship** as ``manifest.json`` — the sealed manifest, nothing dropped.

    Use this rather than a local ``json.dumps``, because the seal covers *which keys exist*
    and a writer is one keyword away from breaking it. ``model_dump(mode="json")`` emits
    ``"schematic": null`` for an M1 bundle (the M2 slot, unfilled), and the digest was taken
    over a payload that had it; a writer using ``exclude_none=True`` — or hand-building a
    "tidier" document — ships a manifest whose own ``manifest_sha256`` no longer matches, and
    the verifier reports the bundle corrupt. Measured: dropping that one null key changes the
    digest completely, and the only symptom is an unusable bundle offline.

    Serialized with JCS as well. The verifier re-canonicalizes whatever it parses, so key
    *order* on the wire is not load-bearing; determinism is worth having anyway, and it means
    the shipped bytes differ from :func:`canonical_manifest_bytes` in exactly one respect —
    the ``integrity`` object is present.
    """
    return rfc8785.dumps(manifest.model_dump(mode="json"))


def build_manifest(
    *,
    bundle_id: str,
    itinerary: FrozenItinerary,
    tiles: TileSourceV1,
    tiles_size_bytes: int,
    walk_graph: Artifact,
    legs: Artifact,
    sites: Artifact,
    narrations: Artifact,
    attribution: Artifact,
    text_license: str | None,
    withheld: Sequence[WithheldValue] = (),
    created_at: datetime | None = None,
) -> BundleManifestV1:
    """Assemble and **seal** the manifest — the compile pipeline's final stage.

    Every argument is already-produced output of an earlier stage; this function invents no
    content. It records the seven per-artifact hashes, totals ``size_bytes``, and seals, so
    the returned manifest's ``integrity.manifest_sha256`` already covers everything below it.

    Args:
        bundle_id: stable id of this compiled bundle (format is the caller's; G12).
        itinerary: :func:`freeze_itinerary`'s output — supplies both ``itinerary_id`` and the
            ``content.itinerary`` artifact, so the two cannot disagree.
        tiles: the **whole** ``TileSourceV1`` from T035, embedded as the card requires (not a
            four-field summary of it); it already carries the archive's own ``sha256``.
        tiles_size_bytes: every byte the tiles stage wrote — the PMTiles archive plus the
            style JSON and glyph directory the ``TileSourceV1`` points at. Those files belong
            to T035/T035b and this module never sees them, so their size is passed in rather
            than guessed: a ``size_bytes`` missing the tiles would under-report the download
            by roughly the whole bundle.
        text_license: ``"CC-BY-SA-4.0"`` when any story is bundled, ``None`` when
            ``content.narrations`` is empty (ADR-0024) — the share-alike declaration, written
            to the card's camelCase ``textLicense``. The narration stage decides it; the
            manifest only records it.
        withheld: what quarantine (T038) removed, with its closed-enum reason. Empty means
            nothing was withheld — never "we did not look".
        created_at: UTC; defaults to now. Explicit only for reproducible builds and tests.

    ``size_bytes`` deliberately excludes the manifest itself: its size is unknown until it is
    sealed (sealing writes a 64-char digest into it), and against a tile archive the few
    hundred bytes are noise. It is a download estimate, not an audit.
    """
    # Two artifacts at one path is not a failure the launch check can explain: the second
    # write wins in the archive, its recorded hash then belongs to bytes that are gone, and
    # the bundle fails verification pointing at the wrong file. Cheap to refuse here.
    _refuse_duplicate_paths(
        tiles.path,
        tiles.style.path,
        tiles.glyphs.path,
        walk_graph.path,
        legs.path,
        sites.path,
        narrations.path,
        itinerary.artifact.path,
        attribution.path,
    )

    size_bytes = tiles_size_bytes + sum(
        artifact.size_bytes
        for artifact in (
            walk_graph,
            legs,
            sites,
            narrations,
            itinerary.artifact,
            attribution,
        )
    )

    fields: dict[str, Any] = {
        "bundle_id": bundle_id,
        "itinerary_id": itinerary.itinerary_id,
        "size_bytes": size_bytes,
        "tiles": BundleTiles(pmtiles=tiles),
        "routing": BundleRouting(
            walk_graph=walk_graph.path,
            walk_graph_sha256=walk_graph.sha256,
            legs=legs.path,
            legs_sha256=legs.sha256,
        ),
        "content": BundleContent(
            sites=sites.path,
            sites_sha256=sites.sha256,
            narrations=narrations.path,
            narrations_sha256=narrations.sha256,
            itinerary=itinerary.artifact.path,
            itinerary_sha256=itinerary.artifact.sha256,
        ),
        "attribution": ArtifactRef(path=attribution.path, sha256=attribution.sha256),
        "textLicense": text_license,
        "withheld": tuple(withheld),
    }
    # Absent rather than `None`: the model's default factory is what supplies "now", and
    # passing None would fail validation instead of falling back to it.
    if created_at is not None:
        fields["created_at"] = created_at

    return BundleManifestV1.seal(**fields)


def _refuse_duplicate_paths(*paths: str) -> None:
    """Refuse two artifacts claiming one bundle path (see :func:`build_manifest`)."""
    seen: set[str] = set()
    duplicates: set[str] = set()
    for path in paths:
        if path in seen:
            duplicates.add(path)
        seen.add(path)
    if duplicates:
        raise ValueError(
            f"two artifacts share a bundle path: {', '.join(sorted(duplicates))}. One path "
            "holds one artifact and one hash; the second write would replace the first."
        )
